core/module.h

```diff
@@ -137,7 +165,43 @@ class ModuleBuilder {
 
   CommandResponse RunInit(Module *m, const google::protobuf::Any &arg) const;
 
+  /*!
+   * Connects two modules together in `module_graph_`.
+   */
+  static int AddEdge(const std::string &from, const std::string &to);
+
+  /*!
+   * Returns task parent(s) of a task node.
+   */
+  static std::unordered_set<Module *> *Parents(const std::string &node) {
+    auto it = task_graph_.find(node);
+    if (it == task_graph_.end()) {
+      return nullptr;
+    }
+    return &(it->second.parents());
+  }
+
  private:
+  /*!
+   * Updates `task_graph_` by traversing `module_graph_` and ignoring all
+   * modules that are not tasks.
+   */
+  static int UpdateTCGraph();
+
+  /*!
+   * Finds the next module that implements a task, and updates `task_graph_`
+   * accordingly.
+   */
+  static int FindNextTC(const std::string &node_name,
+                        const std::string &parent_name,
+                        std::unordered_set<std::string> &visited);
+
+  // A graph of all the modules in the current pipeline.
+  static std::map<std::string, Node> module_graph_;
+
+  // Represents a graph of all modules that are tasks in the pipeline.
+  static std::map<std::string, Node> task_graph_;
```

Comment:
Move the storage of this information to the modules themselves so that they can access their parents via a local member variable (a vector of parents)?